src/envoy/alts/alts_socket_factory.cc

```diff
@@ -29,6 +29,36 @@ namespace Envoy {
 namespace Server {
 namespace Configuration {
 
+using ::google::protobuf::RepeatedPtrField;
+
+// Returns true if the peer's service account is found in targets, otherwise
+// returns false and fills out err with an error message.
+static bool doValidate(const tsi_peer &peer,
+                       const RepeatedPtrField<std::string> &targets,
```

Comment:
pre-process `targets` to a `unordered_set` then you don't have to do O(N^2) validation below.